### ISSUE TITLE
Multiple Crafts listed inside One .craft file?

### ISSUE THREAD

**Issue by Deiscorides** [AJCStriker/Movecraft-3#157]
In order to prevent excessive lists of permissions, I would like to place several craft types inside of 1 craft file. For instance, I would like to have a file named "Gates.craft" and inside it have listed the code for all the different gates. It would basically look like you copy and pasted multiple .craft files into 1 file. This will also help categorize different types of ships and machinery.

When attempting something like this, the plugin only recognizes the very last craft on the list, "name: WoolGate" for instance. It does not matter what order they are in, the last one gets priority and  all other craft types above it are ignored.

I've been looking around for a way to do this, and have not found out how, thus I thought perhaps this feature did not exist yet.

Here is an example of the code that would be inside of 1 .craft file. The dividers are just notation in the file like this # =======

name: AcaciaGate
maxSize: 1000
minSize: 3
allowedBlocks: 
    - "5:4" # Acacia Wood Plank
    - "125:4" # Double Acacia Wood Slab
    - "126:4" # Acacia Wood Slab
    - 162 # Acacia Wood (Logs)
    - 163 # Acacia Wood Stairs
    - 196 # Acacia Door Block
    - 69 # Lever
    - 77 # Stone Button
    - 143 # Wooden Button

forbiddenBlocks:
    - 7 # Bedrock

blockedByWater: false
allowHorizontalMovement: false
allowVerticalMovement: false
speed: 1.0
tryNudge: false
flyblocks:
    85: # Oak Fence
        - 0.0
        - 100.0
# ====================================================================
name: BirchGate
maxSize: 1000
minSize: 3
allowedBlocks: 
    - "5:2" # Birch Wood Plank
    - "17:2" # Birch Wood (Log)
    - "125:2" # Double Birch Wood Slab
    - "126:2" # Birch Wood Slab
    - 135 # Birch Wood Stairs
    - 194 # Birch Door Block
    - 69 # Lever
    - 77 # Stone Button
    - 143 # Wooden Button

forbiddenBlocks:
    - 7 # Bedrock

blockedByWater: false
allowHorizontalMovement: false
allowVerticalMovement: false
speed: 1.0
tryNudge: false
flyblocks:
    85: # Oak Fence
        - 0.0
        - 100.0
# ====================================================================
name: ClayGate
maxSize: 1000
minSize: 3
allowedBlocks: 
    - 82 # Clay
    - 159 # White Stained Clay
    - "159:1" # Orange Stained Clay
    - "159:2" # Magenta Stained Clay
    - "159:3" # Light Blue Stained Clay
    - "159:4" # Yellow Stained Clay
    - "159:5" # Lime Stained Clay
    - "159:6" # Pink Stained Clay
    - "159:7" # Gray Stained Clay
    - "159:8" # Light Gray Stained Clay
    - "159:9" # Cyan Stained Clay
    - "159:10" # Purple Stained Clay
    - "159:11" # Blue Stained Clay
    - "159:12" # Brown Stained Clay
    - "159:13" # Green Stained Clay
    - "159:14" # Red Stained Clay
    - "159:15" # Black Stained Clay
    - 172 # Hardened Clay
    - 69 # Lever
    - 77 # Stone Button
    - 143 # Wooden Button

forbiddenBlocks:
    - 7 # Bedrock

blockedByWater: false
allowHorizontalMovement: false
allowVerticalMovement: false
speed: 1.0
tryNudge: false
flyblocks:
    85: # Oak Fence
        - 0.0
        - 100.0

**Comment by mwkaicz** [comment]
This isn't possible, because *.craft file uses standart YAML structure, you are breaking all rules of making these type of files. Look on it as on Object with keys, values and arrays. You asign key "name" to value "AcaciaGate" and after few rows later you are overwrite it with "BirchGate" ... If you will work with some better IDE, it will tell you, that you can't have two same keys in one Object (.craft file).

There must be something like:
`crafts:`
`    - AcaciaGate:`
`      name: AcaciaGate`
`      minSize: 3` 
`    - BirchGate:`
`      name: BirchGate`
`      minSize: 3` 

... but this isn't supported ... mayby will be better to make a support for subfolders is Movecraft/types

**Comment by Deiscorides** [comment]
Yeah, I know it does not work and I know precisely why it does not work. I not reporting a bug. I'm asking for a feature. Sorry for the confusion.

**Comment by mwkaicz** [comment]
I undersand you very well, but I'm trying to explain you that it's not possible by your way, how you described it above. Only possible ways we have are these two (maybe someone has better one?)

Why you need it all in one file? It can be pretty long file even with one craft inside when you have long blocks lists. Wouldn't be a subfolder better? ... if you need some solution very fast, rename files with some prefixes sou you will see them close together at least. 
But I don't see any usable reason why have more crafts in one file (except of unclear content). I'm not main developer, so final decision will not be mine. But for me, I say "no" for merging more crafts inside one file.

During writing new version of MovecraftView I had idea about modular system. Where you can have make a craft file by merging partial files (like material list file, limits file etc.)  but's it's still only on paper not in code. I would like to start discussion on this topic in January.

**Comment by Deiscorides** [comment]
>Wouldn't be a subfolder better?

If Movecraft allowed use of subfolders inside of the "types" folder, for grouping various types of craft files together, and a single permission could be used for that folder, it'd be great.

I'm just trying to streamline some aspects of permissions. FYI, one way to do this inside 1 file is the same way the Brewery plugin does it for recipes, http://i.imgur.com/3hkyp6D.png by separating them with a number and proper spacing in the yml file.

Regardless, a subfolder and subfolder-permission solution would be best and easiest perhaps to do? It could look something like this as far as folders go. Thus, in this example, only 2 permissions would be needed even if there are 2 or 100 craft files in a subfolder.

```
Types
  Subtype1 (permission set for admin only)
      AdminAirship.craft
      AdminShip.craft
      AdminGate.craft
  Subtype2 (permission set for player(s) only)
      Airship.craft
      Ship.craft
      Gate.craft
```

Modular systems are good too. Using a set of admin-defined modules in one set of files to quickly merge into .craft files would make creating new crafts far easier. Right now, I've been doing a lot of copy-pasting and # notes.

**Comment by AJCStriker** [comment]
I had some interesting thoughts on this when I first designed the .craft system.

It was initially designed to support single parent inheritance that would have allowed for permissions grouping but that was never implemented due to some implementation issues. There are some logical extensions this leads to such as the ability to define one block list and then extend it with specifics.

Note on the semantics of YAML as well: === specifies a new YAML document inside the file according to the specification and could be used to colocate information - if for some reason that was useful.

**Comment by Deiscorides** [comment]
Could the plugin simply have permissions for subfolders in the types folder?

**Comment by mwkaicz** [comment]
Yes, it could. I would like to remake whole file system of this plugin, but it's on long run and it needs more discussion about this topic before I'll start to do something.
